HPCC-28208 Standardize units and bucket labels for scaled histogram m…

kenrowland opened the pull request:
…etrics

Standardized scaled histogram metrics to NS for time based measurements Removed label strings from bucket definitions as the sink will need to generate these anyway.
Added convenience function to create scaled cycles to ns scaled histogram

Signed-Off-By: Kenneth Rowland kenneth.rowland@lexisnexisrisk.com

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-28208 Standardize units and bucket labels for scaled histogram metrics

Standardized scaled histogram metrics to NS for time based measurements
Removed label strings from bucket definitions as the sink will need to
generate these anyway.
Added convenience function to create scaled cycles to ns scaled histogram

Signed-Off-By: Kenneth Rowland kenneth.rowland@lexisnexisrisk.com

ghalliday commented on testing/unittests/metrics/MetricFrameworkTests.cpp:
```diff
@@ -418,6 +429,15 @@ class MetricFrameworkTests : public CppUnit::TestFixture
     }
 
+
+    void checkHistogramValue(std::shared_ptr<IMetric> pHistogram, __uint64 expectedValue, __uint64 error)
```
trivial: Slightly better if 1st parameter is a reference (but doesn't really matter since test code).

ghalliday commented on testing/unittests/metrics/MetricFrameworkTests.cpp:
```diff
@@ -312,16 +312,22 @@ class MetricFrameworkTests : public CppUnit::TestFixture
 
         std::vector<__uint64> values;
 
+        printf("Validating number of returned buckets\n");
```
Possibly better is to use DBGLOG.  If you run with the -v option the logging is traced to stdout

ghalliday commented on system/jlib/jmetrics.hpp:
```diff
@@ -249,17 +249,11 @@ class CustomMetric : public MetricBase
 {
 public:
-    HistogramMetric(const char *name, const char *desc, StatisticMeasure _units, const std::vector<BucketDef> &bucketDefs, const MetricMetaData &_metaData = MetricMetaData());
+    HistogramMetric(const char *name, const char *desc, StatisticMeasure _units, const std::vector<__uint64> &bucketLimitss, const MetricMetaData &_metaData = MetricMetaData());
```
typo: bucketLimitss (I had to find something....).  Second occurrence below

ghalliday reviewed (commented):
Looks good, always nice when the code gets simpler.
A couple of minor comments.

jakesmith commented:
> 1) test: MetricFrameworkTests::Test_histogram_metric (E) 
> uncaught exception of type std::exception (or derived).
> - std::bad_alloc

Looks like this is not as a result of this PR, because other PR's are also failing with same, e.g.:
https://github.com/hpcc-systems/HPCC-Platform/pull/16512

@kenrowland - was there a recent related PR that was merged that is causing this?

kenrowland commented:
@jakesmith @richardkchapman Looking into where it is failing, but cannot see what the new histogram tests could be doing to cause this. They don't allocate any memory other than some small vectors. 

PR https://github.com/hpcc-systems/HPCC-Platform/pull/16463 did not cause any unittest failures

kenrowland commented:
@ghalliday I removed all of the tracing code I added to the unit tests to see if something was going wrong and causing the bad alloc error being seen. Nothing was found to be incorrect. 

I did get one stack dump from Atilla and it seemed it was failing in a base metric class constructor when it was copying the default meta data object. That object defaults to an empty vector and is passed as a const & to parent class constructors. I don't see how it could cause any problems (and I added some output statements to confirm it was a 0 length vector).

I am as confident as I can be that the bad alloc error happening during smoketest/unit tests is not being caused by any errors in the metric framework tests.

ghalliday reviewed (approved):
Looks good - please squash.

kenrowland commented:
@richardkchapman Please merge.
